src/Lean/Elab/Tactic/LibrarySearch.lean

```diff
@@ -24,28 +24,90 @@ Implementation of the `exact?` tactic.
       admitGoal goal
+where
+  /-- Returns all free variables in `e` that are shadowed or not user-enterable. -/
+  collectUnprintableFVars (e : Expr) : MetaM (List FVarId) := do
+    let lctx ← getLCtx
+    let { fvarSet, .. } := Lean.collectFVars {} e
+    let unprintableFVars := fvarSet.toList.filter fun fvarId =>
+      if let some name := lctx.getRoundtrippingUserName? fvarId then
+        name.isInaccessibleUserName || name.isAnonymous || name.hasNum
+      else
+        true
+    return unprintableFVars
+
+  /--
+  Executes `tac` in `savedState` (then restores the current state). Used to ensure that a suggested
+  tactic is valid.
+
+  Remark: we don't merely elaborate the proof term's syntax because it may successfully round-trip
+  (d)elaboration but still produce an invalid tactic (see the example in #5407).
+  -/
+  evalTacticInState (savedState : Tactic.SavedState) (tac : TSyntax `tactic) : TacticM Unit := do
+    let currState ← saveState
+    savedState.restore
+    try
+      evalTactic tac
+    finally
+      currState.restore
+
+  /--
+  Suggests using the value of `goal` as a proof term if the corresponding tactic is valid at
+  `origGoal`, or else informs the user that a proof exists but is not syntactically valid.
+  -/
+  addSuggestionIfValid (ref : Syntax) (goal : MVarId) (initialState : Tactic.SavedState)
+                       (addSubgoalsMsg := false) (errorOnInvalid := true) : TacticM Unit := do
+    let proofExpr := (← instantiateMVars (mkMVar goal)).headBeta
+    let (suggestion, proofMVars) ← mkExactSuggestionSyntax proofExpr
+    let hasMVars := !proofMVars.isEmpty
+    let unprintableFVars ← collectUnprintableFVars proofExpr
+    if unprintableFVars.length > 0 then
+      let exprMessageDatas := unprintableFVars.map (MessageData.ofExpr ∘ Expr.fvar)
+      let fvarsListMessage := MessageData.joinSep exprMessageDatas ", "
+      -- We use `ppExpr` here so that tombstones appear in the output
+      let pp ← withOptions (pp.mvars.set · false) (ppExpr proofExpr)
+      let ppProof := if hasMVars then m!"refine {pp}" else m!"exact {pp}"
+      let msg := m!"found a {if hasMVars then "partial " else ""}proof, but the corresponding tactic\
+                    {indentD ppProof}\n\
+                    is invalid because the following variables have inaccessible names:\
+                    {indentD fvarsListMessage}\n\
+                    To fix this, ensure these variables are not shadowed and are given explicit \
+                    names when introduced."
+      if errorOnInvalid then throwError msg else logInfo msg
+    else
+      try
+        evalTacticInState initialState suggestion
+      catch _ =>
+        let suggestionPretty ← SuggestionText.prettyExtra suggestion
+        let msg := m!"found a {if hasMVars then "partial " else ""}proof, but the corresponding tactic\
+                      {indentD suggestionPretty}\nfailed to compile"
```

Comment:
Somehow “fails to compile” sounds like that’s not the quote the right terminology for tactics, they aren't really compiled, are they? “Fails to elaborate“ maybe? Or simply “failed”?

Reply:
I've opted for "failed," and also moved all the non-tactic text to the top line (this makes the wording a hair less clear but makes it easier to differentiate between the verbiage and the actual tactic).